waiter/src/waiter/state.clj

```diff
@@ -31,44 +31,86 @@
+  (contains? status-tags :expired))
+
 (defn- slots-available?
   "Predicate on healthy instances with available slots greater than used slots."
-  [{:keys [slots-assigned slots-used status-tags]}]
-  (and (healthy? status-tags) (> slots-assigned slots-used)))
+  [{:keys [slots-assigned slots-used] :as state}]
+  (and (healthy? state) (> slots-assigned slots-used)))
+
+(defn- only-lingering-requests?
+  "Returns truthy value if the instance is idle or processing expired requests based on the earliest-request-threshold-time"
+  [request-id->use-reason-map earliest-request-threshold-time]
+  (every? #(t/before? % earliest-request-threshold-time)
+          (->> request-id->use-reason-map
+               vals
+               (map :time))))
 
 (defn- killable?
-  "Only instances without slots used and not killed|locked can be killed."
-  [{:keys [slots-used status-tags]}]
-  (and (zero? slots-used)
-       (not-any? #(contains? status-tags %) [:killed :locked])))
+  "Only the following instances can be killed:
+   - without slots used and not killed|locked, or
+   - expired and idle or processing lingering requests.
+   If there are expired instances, only choose healthy instances that are blacklisted or expired.
+   If there are expired and starting instances, only kill unhealthy instances that are not starting."
+  [request-id->use-reason-map earliest-request-threshold-time expired-instances? starting-instances?
+   {:keys [slots-used status-tags] :as state}]
+  (and (or (and (zero? slots-used)
+                (not-any? #(contains? status-tags %) [:killed :locked]))
+           (and (expired? state)
+                (only-lingering-requests? request-id->use-reason-map earliest-request-threshold-time)))
+       (or (not (and expired-instances? (contains? status-tags :healthy)))
+           (some #(contains? status-tags %) [:blacklisted :expired]))
+       (or (not (and expired-instances? starting-instances?))
+           (and (contains? status-tags :unhealthy)
+                (not (contains? status-tags :starting))))))
+
+(defn- find-max
+  "Uses the provided comparison function to find the max element in collection."
+  [comparison-fn collection]
+  (reduce #(if (pos? (comparison-fn %1 %2)) %1 %2) (first collection) (rest collection)))
+
+(defn- expired-instances?
+  "Returns truthy value if there is some instance in an expired state."
+  [instance-id->state]
+  (some (fn [[_ {:keys [status-tags]}]] (contains? status-tags :expired)) instance-id->state))
+
+(defn- starting-instances?
+  "Returns truthy value if there is some instance in an starting state."
+  [instance-id->state]
+  (some (fn [[_ {:keys [status-tags]}]] (contains? status-tags :starting)) instance-id->state))
 
 (defn find-killable-instance
   "While killing instances choose using the following logic:
    if there are expired and starting instances, only kill unhealthy instances that are not starting;
    choose instances in following order:
    - choose amongst idle instances
```

Comment:
The method documentation needs an overhaul given the new logic.

Reply:
Done.